torchtune/models/llama3/_tokenizer.py

```diff
@@ -0,0 +1,175 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, List, Optional, Tuple
+
+from torchtune.data import Message, truncate
+from torchtune.modules.tokenizers import ModelTokenizer, TikTokenBaseTokenizer
+
+
+CL100K_PATTERN = r"""(?i:'s|'t|'re|'ve|'m|'ll|'d)|[^\r\n\p{L}\p{N}]?\p{L}+|\p{N}{1,3}| ?[^\s\p{L}\p{N}]+[\r\n]*|\s*[\r\n]+|\s+(?!\S)|\s+"""  # noqa
+
+PAD_ID = 0
+
+LLAMA3_SPECIAL_TOKENS = {
+    "<|begin_of_text|>": 128000,
+    "<|end_of_text|>": 128001,
+    "<|start_header_id|>": 128006,
+    "<|end_header_id|>": 128007,
+    "<|eot_id|>": 128009,
+    "<|eom_id|>": 128008,
+    "<|python_tag|>": 128255,
+}
+
+
+class Llama3Tokenizer(ModelTokenizer):
+    """
+    tiktoken tokenizer configured with Llama3 Instruct's special tokens, as described in
+    https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3
+
+    Args:
+        path (str): Path to pretrained tiktoken tokenizer file.
+        special_tokens (Optional[Dict[str, int]]): mapping containing special text tokens and
+            their registered token IDs. If left as None, this will be set to the canonical
+            Llama3 special tokens.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        special_tokens: Optional[Dict[str, int]] = None,
```

Comment:
Why wouldn't we just default this to the LLAMA3_SPECIAL_TOKENS?

Reply:
mutable types - if we put dict/list as default here it will be the same dict for all instances of Llama3Tokenizer. If it's ever updated it will updated for all of them. Very edge case, but better to just be safe.